src/Symfony/Component/Messenger/Middleware/WrapsMessageHandlingInTransaction.php

```diff
@@ -0,0 +1,65 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Messenger\Middleware;
+
+use Doctrine\Common\Persistence\ManagerRegistry;
+use Doctrine\ORM\EntityManagerInterface;
+use Symfony\Component\Messenger\MiddlewareInterface;
+
+/**
+ * Wraps all handlers in a single doctrine transaction.
+ *
+ * @author Tobias Nyholm <tobias.nyholm@gmail.com>
+ */
+class WrapsMessageHandlingInTransaction implements MiddlewareInterface
+{
+    /**
+     * @var ManagerRegistry
+     */
+    private $managerRegistry;
+
+    /**
+     * @var string
+     */
+    private $entityManagerName;
+
+    /**
+     * @param ManagerRegistry $managerRegistry
+     * @param string          $entityManagerName
+     */
+    public function __construct(ManagerRegistry $managerRegistry, string $entityManagerName)
+    {
+        $this->managerRegistry = $managerRegistry;
+        $this->entityManagerName = $entityManagerName;
+    }
+
+    public function handle($message, callable $next)
+    {
+        /** @var $entityManager EntityManagerInterface */
+        $entityManager = $this->managerRegistry->getManager($this->entityManagerName);
+
+        $result = null;
+        try {
+            $entityManager->transactional(
```

Comment:
It looks like to me that using the `beginTransaction`, `commit` and `rollback` methods would be much more explicit that this callback-based transaction that forces you to do this weird `resetManager` later :)